crates/ra_assists/src/handlers/add_function.rs

```diff
@@ -0,0 +1,720 @@
+        fn_buf.push_str(&self.fn_generics);
+        fn_buf.push_str(&self.fn_args);
+        fn_buf.push_str(" {\n");
+        fn_buf.push_str(&self.indent);
+        fn_buf.push_str("    ");
+
+        // We take the offset here to put the cursor in front of the `todo` body
+        let offset = TextUnit::of_str(&fn_buf);
+
+        fn_buf.push_str("todo!()\n");
+        fn_buf.push_str(&self.indent);
+        fn_buf.push_str("}");
+
+        let cursor_pos = self.start_offset + offset;
+        FunctionTemplate {
+            fn_text: fn_buf,
+            cursor_offset: cursor_pos,
+            insert_offset: self.start_offset,
+        }
+    }
+}
+
+fn fn_name(call: &CallExpr) -> Option<String> {
+    Some(call.expr()?.syntax().to_string())
+}
+
+fn fn_generics(_call: &CallExpr) -> Option<String> {
+    // TODO
+    Some("".into())
+}
+
+fn fn_args(ctx: &AssistCtx, call: &CallExpr) -> Option<String> {
+    let mut arg_names = Vec::new();
+    let mut arg_types = Vec::new();
+    for arg in call.arg_list()?.args() {
+        dbg!(&arg);
+        let arg_name = match fn_arg_name(&arg) {
+            Some(name) => name,
+            None => String::from("arg"),
+        };
+        arg_names.push(arg_name);
+        arg_types.push(match fn_arg_type(ctx, &arg) {
+            Some(ty) => ty,
+            None => String::from("()"),
+        });
+    }
+    deduplicate_arg_names(&mut arg_names);
+    Some(format!(
+        "({})",
+        arg_names
+            .into_iter()
+            .zip(arg_types)
+            .map(|(name, ty)| format!("{}: {}", name, ty))
+            .collect::<Vec<_>>()
+            .join(", ")
+    ))
+}
+
+/// Makes duplicate argument names unique by appending incrementing numbers.
+///
+/// ```
+/// let mut names: Vec<String> =
+///     vec!["foo".into(), "foo".into(), "bar".into(), "baz".into(), "bar".into()];
+/// deduplicate_arg_names(&mut names);
+/// let expected: Vec<String> =
+///     vec!["foo_1".into(), "foo_2".into(), "bar_1".into(), "baz".into(), "bar_2".into()];
+/// assert_eq!(names, expected);
+/// ```
+fn deduplicate_arg_names(arg_names: &mut Vec<String>) {
+    let arg_name_counts = arg_names.iter().fold(FxHashMap::default(), |mut m, name| {
+        *m.entry(name).or_insert(0) += 1;
+        m
+    });
+    let duplicate_arg_names: FxHashSet<String> = arg_name_counts
+        .into_iter()
+        .filter(|(_, count)| *count >= 2)
+        .map(|(name, _)| name.clone())
+        .collect();
+
+    let mut counter_per_name = FxHashMap::default();
+    for arg_name in arg_names.iter_mut() {
+        if duplicate_arg_names.contains(arg_name) {
+            let counter = counter_per_name.entry(arg_name.clone()).or_insert(1);
+            arg_name.push('_');
+            arg_name.push_str(&counter.to_string());
+            *counter += 1;
+        }
+    }
+}
+
+fn fn_arg_name(fn_arg: &Expr) -> Option<String> {
+    if let Some(cast_expr) = ast::CastExpr::cast(fn_arg.syntax().clone()) {
```

Comment:
Without this, `x as u8` will get the name `u8` instead of `x`.
I'm curious how many special cases all those different kinds of expressions will require...